website/blog/2025/01-29-superscraper/index.md

```diff
@@ -0,0 +1,189 @@
+The project uses Node.js `http` module to create a server that listens on the desired port. After the server starts, a check ensures users are interacting with it correctly by sending requests instead of running it traditionally. This keeps SuperScraper operating as a persistent server.
+
+### Handling multiple crawlers
+
+SuperScraper processes user requests using multiple instances of Crawlee’s `PlaywrightCrawler`. Since each `PlaywrightCrawler` can only handle one proxy configuration, a separate crawler is created for each unique proxy setting. 
+
+For example, if the user sends one request for “normal” proxies and one request with residential US proxies, a separate crawler needs to be created for each proxy configuration. Hence, to solve this, we store the crawlers in a key-value map, where the key is a stringified proxy configuration.
+
+```js
+const crawlers = new Map<string, PlaywrightCrawler>();
+```
+
+Here’s a part of the code that gets executed when a new request from the user arrives; if the crawler for this proxy configuration exists in the map, it will be used. Otherwise, a new crawler gets created. Then, we add the request to the crawler’s queue so it can be processed.
+
+```js
+const key = JSON.stringify(crawlerOptions); 
+const crawler = crawlers.has(key) ? crawlers.get(key)! : await createAndStartCrawler(crawlerOptions);
+
+await crawler.requestQueue!.addRequest(request);
+```
+
+The function below initializes new crawlers with predefined settings and behaviors. Each crawler utilizes its own in-memory queue created with the `MemoryStorage` client. This approach is used for two key reasons:
+
+1. **Performance**: In-memory queues are faster, and there's no need to persist them when SuperScraper migrates.
+2. **Isolation**: Using a separate queue prevents interference with the shared default queue of the SuperScraper Actor, avoiding potential bugs when multiple crawlers use it simultaneously.
+
+```js
+export const createAndStartCrawler = async (crawlerOptions: CrawlerOptions = DEFAULT_CRAWLER_OPTIONS) => {
+    const client = new MemoryStorage();
+    const queue = await RequestQueue.open(undefined, { storageClient: client });
+
+    const proxyConfig = await Actor.createProxyConfiguration(crawlerOptions.proxyConfigurationOptions);
+
+    const crawler = new PlaywrightCrawler({
+        keepAlive: true,
+        proxyConfiguration: proxyConfig,
+        maxRequestRetries: 4,
+        requestQueue: queue,
+    });
+};
+```
+
+At the end of the function, we start the crawler and log a message if it terminates for any reason. Next, we add the newly created crawler to the key-value map containing all crawlers, and finally, we return the crawler.
+
+```js
+crawler.run().then(
+    () => log.warning(`Crawler ended`, crawlerOptions),
+    () => { }
+);
+
+crawlers.set(JSON.stringify(crawlerOptions), crawler);
+
+log.info('Crawler ready 🚀', crawlerOptions);
+
+return crawler;
+```
+
+### Mapping standby HTTP requests to Crawlee requests
```

Comment:
These are just isolated function and you need to browse the repo (or use a lot of imagination) to see how it all fits together. I'd appreciate a more top-down approach - maybe start with a code snippet with the big picture and then show how the individual functions are implemented?